test/contrib/spark_test.py

```diff
@@ -15,19 +15,119 @@
 # limitations under the License.
 #
 
-try:
-    from StringIO import StringIO
-except ImportError:
-    from io import StringIO
-import subprocess
 from helpers import unittest
-
 import luigi
 import luigi.hdfs
-from helpers import with_config
-from luigi.contrib.spark import PySpark1xJob, Spark1xJob, SparkJob, SparkJobError
+import subprocess
+from luigi import six
 from luigi.mock import MockTarget
+from helpers import with_config
+from luigi.contrib.spark import SparkJobError, SparkSubmitTask, PySpark1xJob, Spark1xJob, SparkJob
 from mock import patch
+StringIO = six.BytesIO
+
+
+class TestSparkSubmitTask(SparkSubmitTask):
+    deploy_mode = "client"
+    name = "AppName"
+    entry_class = "org.test.MyClass"
+    jars = ["jars/my.jar"]
+    py_files = ["file1.py", "file2.py"]
+    files = ["file1", "file2"]
+    conf = {"Prop": "Value"}
+    properties_file = "conf/spark-defaults.conf"
+    driver_memory = "4G"
+    driver_java_options = "-Xopt"
+    driver_library_path = "library/path"
+    driver_class_path = "class/path"
+    executor_memory = "8G"
+    driver_cores = 8
+    supervise = True
+    total_executor_cores = 150
+    executor_cores = 10
+    queue = "queue"
+    num_executors = 2
+    archives = ["archive1", "archive2"]
+    app = "file"
+
+    def app_options(self):
+        return ["arg1", "arg2"]
+
+    def output(self):
+        return luigi.LocalTarget('output')
+
+
+class TestDefaultSparkSubmitTask(SparkSubmitTask):
+    app = 'test.py'
+
+    def output(self):
+        return luigi.LocalTarget('output')
+
+
+class SparkSubmitTest(unittest.TestCase):
+    ss = 'ss-stub'
+
+    @with_config({'spark': {'spark-submit': ss, 'master': "local[*]"}})
+    @patch('luigi.contrib.spark.subprocess.Popen')
+    def test_run(self, mock):
+        mock.return_value.returncode = 0
+
+        def poll_generator():
+            yield None
+            yield 1
+
+        mock.return_value.poll = (lambda gen: lambda: gen.next() if hasattr(gen, 'next') else gen.__next__())(poll_generator())
```

Comment:
Something like this would probably be clearer and work with both python2 and 3

```python
poll_gen = poll_generator()
mock.return_value.poll = lambda: next(poll_gen)
```